src/py/flwr/server/superlink/state/in_memory_state.py

```diff
@@ -205,24 +210,39 @@ def num_task_res(self) -> int:
         """
         return len(self.task_res_store)
 
-    def create_node(self, ping_interval: float) -> int:
+    def create_node(
+        self, ping_interval: float, public_key: Optional[bytes] = None
+    ) -> int:
         """Create, store in state, and return `node_id`."""
         # Sample a random int64 as node_id
         node_id: int = int.from_bytes(os.urandom(8), "little", signed=True)
 
         with self.lock:
             if node_id not in self.node_ids:
                 self.node_ids[node_id] = (time.time() + ping_interval, ping_interval)
+                if (
+                    public_key is not None
+                    and public_key not in self.public_key_to_node_id
+                    and self.public_key_to_node_id.get(public_key) != node_id
+                ):
+                    self.public_key_to_node_id[public_key] = node_id
                 return node_id
         log(ERROR, "Unexpected node registration failure.")
         return 0
 
-    def delete_node(self, node_id: int) -> None:
+    def delete_node(self, node_id: int, public_key: Optional[bytes] = None) -> None:
         """Delete a client node."""
         with self.lock:
             if node_id not in self.node_ids:
                 raise ValueError(f"Node {node_id} not found")
             del self.node_ids[node_id]
+            if public_key is not None:
+                if (
+                    public_key not in self.public_key_to_node_id
+                    and self.public_key_to_node_id.get(public_key) != node_id
+                ):
+                    raise ValueError("Public key or node_id not found")
+                del self.public_key_to_node_id[public_key]
```

Comment:
This is similar to the comment above: if both `node_id` and `public_key` are provided, we must ensure that either both `node_ids` and `public_key_to_node_id ` are modified, or none of them is modified.

Reply:
Move node_id deletion after public_key check.